src/measurement/measurement_functions.rs

```diff
@@ -0,0 +1,227 @@
+// SPDX-License-Identifier: Apache-2.0
+
+//! Operations to calculate guest measurement for different SEV modes
+use crate::measurement::{
+    gctx::Gctx,
+    ovmf::{OvmfSevMetadataSectionDesc, SectionType, OVMF},
+    sev_hashes::SevHashes,
+    vcpu_types::CpuType,
+    vmsa::{SevMode, VMMType, VMSA},
+};
+use hex::FromHex;
+use std::path::PathBuf;
+use std::str::FromStr;
+
+use crate::error::*;
+
+use openssl::sha::Sha256;
+
+const _PAGE_MASK: u64 = 0xfff;
+
+/// Get the launch digest as a hex string
+pub fn get_hex_ld(ld: Vec<u8>) -> String {
+    hex::encode(ld)
+}
+
+/// Update launch digest with SEV kernel hashes
+fn snp_update_kernel_hashes(
+    gctx: &mut Gctx,
+    ovmf: &OVMF,
+    sev_hashes: Option<&SevHashes>,
+    gpa: u64,
+    size: usize,
+) -> Result<(), MeasurementError> {
+    match sev_hashes {
+        Some(hash) => {
+            let sev_hashes_table_gpa = ovmf.sev_hashes_table_gpa()?;
+            let page_offset = sev_hashes_table_gpa & _PAGE_MASK;
+            let sev_hashes_page = hash.construct_page(page_offset as usize)?;
+            assert_eq!(sev_hashes_page.len(), size);
+            gctx.update_normal_pages(gpa, sev_hashes_page.as_slice())?
+        }
+        None => gctx.update_zero_pages(gpa, size)?,
+    }
+
+    Ok(())
+}
+
+/// Update launch digest with different section types
+fn snp_update_section(
+    desc: &OvmfSevMetadataSectionDesc,
+    gctx: &mut Gctx,
+    ovmf: &OVMF,
+    sev_hashes: Option<&SevHashes>,
+    vmm_type: VMMType,
+) -> Result<(), MeasurementError> {
+    match desc.section_type {
+        SectionType::SnpSecMemory => gctx.update_zero_pages(desc.gpa.into(), desc.size as usize)?,
+        SectionType::SnpSecrets => gctx.update_secrets_page(desc.gpa.into())?,
+        SectionType::CPUID => {
+            if vmm_type != VMMType::EC2 {
+                gctx.update_cpuid_page(desc.gpa.into())?
+            }
+        }
+        SectionType::SnpKernelHashes => {
+            snp_update_kernel_hashes(gctx, ovmf, sev_hashes, desc.gpa.into(), desc.size as usize)?
```

Comment:
Shouldn't `sev_hashes` not be an `Option` at this point?

Reply:
What doe you mean?

Reply:
Never mind, I now see that you update a zero page if `sev_hashes` is `None`.